Improve service discovery and support IPv6

wutschel opened the pull request:
## Description
<!--- Detailed info for reviewers and developers -->
Closes https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/256.

This PR improves the Remote App's capability during service discovery (when pressing "Find Kodi"):
- Enable support for IPv6 and localhost
- Resolve TCP port automatically
- User setting for preferred address type (IPv4, IPv6 or localhost)
- Bring back "Find Kodi" button on service discovery error

## Summary for release notes
<!--- This will be shown in Testflight to end users / testers -->
<!--- Please fill in, as usually PR title isn't clear to ordinary users -->
<!--- If your changes don't modify app files, please add prefix [not app] -->
Improvement: Support IPv6 and localhost during service discovery
Improvement: Resolve TCP port during service discovery
Improvement: User selectable address type preference
Bugfix: Re-enable "Find Kodi" on service discovery error

kambala-decapitator commented on XBMC Remote/HostViewController.m:
```diff
@@ -311,37 +311,69 @@ - (void)resolveIPAddress:(NSNetService*)service {
 
 - (void)netServiceDidResolveAddress:(NSNetService*)service {
-
+    NSMutableDictionary *serverAdresses = [NSMutableDictionary new];
```
serverAd**d**resses :)

wutschel replied:
German :)

kambala-decapitator commented on XBMC Remote/HostViewController.m:
```diff
@@ -204,17 +205,27 @@ - (void)netServiceBrowser:(NSNetServiceBrowser*)browser didNotSearch:(NSDictiona
+            [self stopDiscovery];
+        }
+    }
+    else if ([name isEqualToString:descriptionUI.text]) {
+        [self resolveIPAddress:aNetService];
```
could you explain why this is needed?

wutschel replied:
This is the call to find the TCP port. Calling `resolveIPAddress` triggers `[remoteService resolveWithTimeout:0]` which finally ends up in `netServiceDidResolveAddress` being called. In `netServiceDidResolveAddress` the port is identified. Is there a different way to achieve this?

kambala-decapitator replied:
Thanks a lot. I simply didn't understand the purpose at first :)

wutschel replied:
Then I better add a proper comment. :)

kambala-decapitator commented on XBMC Remote/HostViewController.m:
```diff
@@ -307,50 +318,75 @@ - (void)fillMacAddressInfo {
+        }
+        NSLog(@"Resolved address/port for '%@': %@", service.name, serverAdresses);
+    }
+    else {
```
probably worth mentioning the checked case explicitly

wutschel replied:
What case do you miss? '_xbmc-jsonrpc-h._tcp.'?

kambala-decapitator replied:
yep

kambala-decapitator commented on XBMC Remote/GlobalData.h:
```diff
@@ -12,17 +12,19 @@
     NSString *serverUser;
     NSString *serverPass;
-    NSString *serverIP;
+    NSString *serverRawIP; // Holds IP in raw format, e.g. ":::::" for IPv6
```
do you really need a variable for that? Couldn't you just make it a "computed property" (readonly or simply a method)?

wutschel replied:
Of course I could, I just wanted to avoid to touch too much code ... Let me look into this.

kambala-decapitator replied:
I mean, you'd still have the `serverRawIP` property, just have a `- (NSString *)serverRawIP` method to build it instead of storing the result

wutschel replied:
It's the other way round. The variable `serverRawIP` holds the unaltered address, and `serverIP` holds the one which has the URL-style variant. I can of course remove one of both and replace all accesses to `serverIP` by `getUrlStyleAddress`. I just feel this is adding quite some overhead as this would be called many times.

kambala-decapitator replied:
right, messed them up. Ok, let's keep then.

kambala-decapitator commented on XBMC Remote/HostViewController.m:
```diff
@@ -295,60 +298,146 @@ - (NSString*)resolveMacFromIP:(NSString*)ipAddress {
+            char addressBuffer[MAX(INET_ADDRSTRLEN, INET6_ADDRSTRLEN)];
+            struct sockaddr_in *socketAddress = (struct sockaddr_in*)[data bytes];
+            int sockFamily = socketAddress->sin_family;
+            if (sockFamily == AF_INET) {
```
please use `switch`

wutschel replied:
ok, even though I feel this makes the code harder to read in this case.

One question: What is the condition to assign `serverAddresses[@"hostname"]`? Currently it is only assigned in case of `AF_INET`, and only if _both_ IPv4 address string and `port` are non-empty.

kambala-decapitator replied:
if it's really worse, then let's keep `if`

hostname: I believe any case should be fine. AFAIK on my LAN it's normally resolved as IPv6, but it shouldn't matter actually.

wutschel replied:
Change is already done, I'll provide a fix commit and you can judge.

wutschel replied:
> hostname: I believe any case should be fine. AFAIK on my LAN it's normally resolved as IPv6, but it shouldn't matter actually.

But can there be situations where hostname can be found, but it will not enter the `AF_INET` case, or there is no `addressStr`? I could for example add the assignment for each of the AF_INET and AF_INET6 cases, but outside of the `addressStr` condition.

kambala-decapitator replied:
oh yes, makes sense, although I'm not sure that a case without address is possible. Anyway, let's first check only for port and add hostname, then also check for the address.

wutschel replied:
Done

kambala-decapitator reviewed (changes requested):
in general seems fine, haven't tested yet

btw commit message "Handle error while service discovery" feels like a few words are missing :)

wutschel commented on XBMC Remote/HostViewController.m:
```diff
@@ -296,18 +285,19 @@ - (NSString*)resolveMacFromIP:(NSString*)ipAddress {
         mac_1_UI.text = macPart[1];
-        mac_1_UI.textColor = [Utilities getSystemBlue];
         mac_2_UI.text = macPart[2];
-        mac_2_UI.textColor = [Utilities getSystemBlue];
         mac_3_UI.text = macPart[3];
-        mac_3_UI.textColor = [Utilities getSystemBlue];
         mac_4_UI.text = macPart[4];
-        mac_4_UI.textColor = [Utilities getSystemBlue];
         mac_5_UI.text = macPart[5];
+        
+        mac_0_UI.textColor = [Utilities getSystemBlue];
+        mac_1_UI.textColor = [Utilities getSystemBlue];
+        mac_2_UI.textColor = [Utilities getSystemBlue];
+        mac_3_UI.textColor = [Utilities getSystemBlue];
+        mac_4_UI.textColor = [Utilities getSystemBlue];
         mac_5_UI.textColor = [Utilities getSystemBlue];
```
Added the label to a NSArray and looping through it now.

wutschel commented on XBMC Remote/HostViewController.m:
```diff
@@ -311,37 +311,69 @@ - (void)resolveIPAddress:(NSNetService*)service {
+                    @"addr": [NSString stringWithFormat:@"%s", addressStr],
+                    @"port": [NSString stringWithFormat:@"%d", port],
+                };
+                serverAdresses[@"localhost"] = @{
```
Required? In this case I would need to change also the setting value.

wutschel replied:
Was thinking one it. Shall also the setting then be named "host name"?

kambala-decapitator replied:
yes, of course, thanks for noticing

kambala-decapitator commented on XBMC Remote/HostViewController.m:
```diff
@@ -299,19 +299,19 @@ - (void)fillMacAddressInfo {
+                mac_4_UI,
+                mac_5_UI,
+            ];
+        for (int i = 0; i < macLabels.count; ++i) {
```
FYI the preferred way to loop through dic/array is `[obj enumerate...]`, but not insisting on changing

wutschel replied:
Ok, took this as opportunity for learning. Hopefully I did this right.

kambala-decapitator commented on XBMC Remote/HostViewController.m:
```diff
@@ -331,36 +331,42 @@ - (void)netServiceDidResolveAddress:(NSNetService*)service {
-                }
+            switch (sockFamily) {
+                case AF_INET: {
+                        const char *addressStr = inet_ntop(sockFamily,
```
not sure why the indent is 2x, is it Xcode's auto formatting?

wutschel replied:
I did this. Having the closing "}" indented less than "break" I find confusing. I can change this back, if you prefer this. Or we just keep the if-else :)

kambala-decapitator replied:
let's revert to if :)

wutschel replied:
Done

kambala-decapitator commented on XBMC Remote/HostViewController.m:
```diff
@@ -299,19 +299,19 @@ - (void)fillMacAddressInfo {
-        mac_4_UI.textColor = [Utilities getSystemBlue];
-        mac_5_UI.textColor = [Utilities getSystemBlue];
+        NSArray *macLabels = @[
+                mac_0_UI,
+                mac_1_UI,
+                mac_2_UI,
+                mac_3_UI,
+                mac_4_UI,
+                mac_5_UI,
+            ];
```
please remove 1 indent

wutschel replied:
Done

kambala-decapitator commented on XBMC Remote/HostViewController.m:
```diff
@@ -339,15 +339,17 @@ - (void)netServiceDidResolveAddress:(NSNetService*)service {
+                            if (addressStr) {
+                                serverAddresses[@"ipv4"] = @{
+                                    @"addr": [NSString stringWithFormat:@"%s", addressStr],
+                                    @"port": [NSString stringWithFormat:@"%d", port],
```
please extract port's string to a variable to prevent copy-paste

wutschel replied:
Did this, and added some helper methods to reduce the code duplication further.

kambala-decapitator commented on XBMC Remote/HostViewController.m:
```diff
@@ -308,11 +308,11 @@ - (void)fillMacAddressInfo {
-        for (int i = 0; i < macLabels.count; ++i) {
-            UILabel *label = macLabels[i];
-            label.text = macPart[i];
+        [macLabels enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
+            UILabel *label = obj;
```
well done! but it can be simplified a bit:

```suggestion
        [macLabels enumerateObjectsUsingBlock:^(UILabel *label, NSUInteger idx, BOOL *stop) {
```

wutschel replied:
Nice!

kambala-decapitator commented on XBMC Remote/HostViewController.m:
```diff
@@ -337,39 +361,19 @@ - (void)netServiceDidResolveAddress:(NSNetService*)service {
-                        };
-                    }
-                }
+                NSString *addr = [NSString stringWithFormat:@"%s", addressStr];
```
when `addressStr == NULL`, you'll receive `"(null)"`, not empty string.

also, it's easier to write `@(addressStr)`, but make sure not to do that on NULL.

wutschel replied:
Done

kambala-decapitator commented on XBMC Remote/HostViewController.m:
```diff
@@ -337,39 +361,19 @@ - (void)netServiceDidResolveAddress:(NSNetService*)service {
-                    }
-                }
+                NSString *addr = [NSString stringWithFormat:@"%s", addressStr];
+                NSString *port = [NSString stringWithFormat:@"%d", ntohs(socketAddress->sin_port)];
```
feels easier to pass raw port value to the function

wutschel replied:
Done

wutschel commented:
That was hard to update. I now added the changes step-wise again. Comparison to former content shows two additional changes:
- Added string "Host name" to translation
- Do not add a new line to tcpJSONRPC.m

This is rebased to master and good to go.

kambala-decapitator commented:
> Just reading through this again: Shouldn't we check for "if (port > 0)" instead if "If (port)"?

Yes, it's slightly better, although I don't believe you can receive negative port in our case

wutschel commented:
> Yes, it's slightly better, although I don't believe you can receive negative port in our case

Chaged this to `if (port >0)` in both cases.

kambala-decapitator commented:
just tested TF: TCP port isn't resolved :( any idea? I use custom port.

wutschel commented:
> just tested TF: TCP port isn't resolved :( any idea? I use custom port.

Yes, I wrote you via PM on this. The TCP port recognition only works for IPv4. I was searching the web for how to resolve with IPv6, but could not find a solution which I was able to implement.

kambala-decapitator commented:
> > just tested TF: TCP port isn't resolved :( any idea? I use custom port.
> 
> Yes, I wrote you via PM on this. The TCP port recognition only works for IPv4. I was searching the web for how to resolve with IPv6, but could not find a solution which I was able to implement.

oh right, sorry, I forgot. Thanks a lot for reminding!

wutschel commented:
> oh right, sorry, I forgot. Thanks a lot for reminding!

Wait, I was responding too quick. I wrote you on the MAC address, the TCP port should be resolved.
